HACKING.jst.md

```diff
@@ -0,0 +1,100 @@
+  
+* `parser-standard` is meant to be very similar to the parser used in the
+  OCaml compiler. Its role in `ocamlformat` is (only) to provide a safety-check:
+  we want to make sure that the formatting does not change the AST as parsed by
+  the compiler. So after formatting, the parsed AST is checked against the
+  original parsed AST to make sure they are the same (modulo some normalization,
+  as written in `lib/Normalize_std_ast.ml`). Key point: this parser and its
+  representation are *not* pretty-printed and never interact with the
+  pretty-printer. The *only* reason this is here is to mimic the behavior of the
+  OCaml compiler. Accordingly, it should be as close to the OCaml compiler's
+  parser as possible.
+  
+* `parser-extended` uses an extended parsetree, capable of storing extra
+  information useful in preserving the structure of the user's code. It is this
+  extended parsetree that gets pretty-printed. The parser here forms part of the
+  core implementation of `ocamlformat`.
+  
+* `parser-recovery` is used for partial parsing, so that a bogus input source
+  file does not get mangled. It was an experiment that has been discontinued by
+  upstream and is not used with Jane Street. It uses the same parsetree as
+  `parser-extended`. A patchfile tracking the changes between `parser-extended`
+  and `parser-recovery` is generated. Just accept any changes that accrue there.
+  
+The directory `vendor/ocaml-common` contains files that are shared between
+`parser-standard` and `parser-extended`, like `location.ml`.
+  
+Testing is done in the `test` directory. TODO: Write more about the testing
+setup.
+
+Design considerations
+---------------------
+
+Because the value of `parser-standard` is entirely in its ability to mimic the
+compiler's parser, we want to keep this parser as close as possible to the
+compiler's. We will want to copy over any changes made to the compiler's parser
+into this version of `ocamlformat`'s parser.
+
+On the other hand, the `parser-extended` can go off in its own direction: its
+parsetree should be designed to make pretty-printing easy. In addition, we want
+to make sure that incorporating upstream changes is easy. We thus feel free to
+edit the parsetree in `parser-extended`, but we do so by adding new
+constructors, not modifying existing ones. In addition, new code should be
+marked off with comments like `(* Jane Street extension *)` and `(* End Jane
+Street extension *)`. Because of the ability to extend the parsetree here, we do
+*not* use jane-syntax in `parser-extended`.
+
+How to update `ocamlformat`
+---------------------------
+
+The base branch to work from is called `jane`. Create a branch off of `jane`.
+
+1. Take the patch you wish to support (i.e. some PR in `flambda-backend`).
+   Apply any changes to the `ocaml/parsing` directory to the files in
+   `vendor/parser-standard`. Remember: this "standard" parser should be as
+   close as possible to the compiler's.
```

Comment:
Also, note here that some files used by both parsers are stored in `vendor/ocaml-common` and may need to be updated.  Further, when incorporating new support files from the compiler, consider whether than can be shared in that directory rather than copied into each of the parser directories.  This is typically the case if the support module doesn't depend on the parsetree.